lib/syslog/writer.go

```diff
@@ -241,18 +244,27 @@ func (c bufferedConn) Close() error                { return c.conn.Close() }
 func (c bufferedConn) Flush() error                { return c.buf.Flush() }
 func (c bufferedConn) Write(b []byte) (int, error) { return c.buf.Write(b) }
 
-func dialWriter(network string, address string, config *tls.Config) (w io.Writer, err error) {
+func dialWriter(network string, address string, config *tls.Config, socksProxy string) (w io.Writer, err error) {
 	var conn net.Conn
 	var dial func(string, string) (net.Conn, error)
+	var socksDialer *SocksDialer
 
 	if network == "tls" {
-		network, dial = "tcp", func(network string, address string) (net.Conn, error) {
+		network, dial = "tcp", func(network, address string) (net.Conn, error) {
 			return tls.Dial(network, address, config)
 		}
 	} else {
 		dial = net.Dial
 	}
 
+	if socksProxy != "" {
+		socksDialer = &SocksDialer{tlsConfig: config}
+		if err = socksDialer.connect(network, socksProxy); err != nil {
+			return
+		}
```

Comment:
Is `connect` just constructing the `dialer` in the `SocksDialer`? If so I would suggest two different approaches because the naming is a little weird:

- Change this to be a constructor, something like `NewSocksDialer(*tls.Config) (*SocksDialer, error)`, to make the `SocksDialer` type immutable. Here's it's unclear whether `connect` can be called multiple times.

- Get rid of the `SocksDialer` type entirely and just use `proxy.SOCKS5` and set the `dial` variable, something like this:
```go
if d, err = proxy.SOCKS5(network, socksProxy, nil, proxy.Direct); err != nil {
  return
}
dial = func(network, address string) (conn net.Conn, err error) {
  // put the content of (*SocksDialer).dial here
}
```